src/middlewared/middlewared/plugins/auth_/2fa.py

```diff
@@ -88,23 +89,74 @@ async def do_update(self, data):
-            'datastore.query', 'account.twofactor_user_auth', [['user', '=', user_id]], {'get': True}
-        )
+    async def get_user_config(self, user_id, local_user):
+        filters = [
+            ['user', '=', user_id], ['user_sid', '=', None]
+        ] if local_user else [['user_sid', '=', user_id], ['user', '=', None]]
+        if config := await self.middleware.call('datastore.query', 'account.twofactor_user_auth', filters):
+            return {
+                **config[0],
+                'exists': True,
+            }
+        else:
+            return {
+                'secret': None,
+                filters[0][0]: user_id,
+                'exists': False,
+            }
 
     @private
     def generate_base32_secret(self):
         return pyotp.random_base32()
 
     @private
     def get_users_config(self):
-        return [
-            {
-                'username': config['user']['bsdusr_username'],
-                'secret_hex': base64.b16encode(base64.b32decode(config['secret'])).decode()
-            }
-            for config in self.middleware.call_sync(
-                'datastore.query', 'account.twofactor_user_auth', [['secret', '!=', None]]
+        users = []
+        mapping = {
+            user['sid']: user for user in self.middleware.call_sync(
+                'user.query', [['local', '=', False], ['sid', '!=', None]], {
+                    'extra': {'additional_information': ['DS']},
+                }
+            )
+        }
+        for config in self.middleware.call_sync(
+            'datastore.query', 'account.twofactor_user_auth', [['secret', '!=', None]]
+        ):
+            username = None
+            if config['user']:
+                username = config['user']['bsdusr_username']
+            elif user := mapping.get(config['user_sid']):
+                username = user['username']
+
+            if username:
+                users.append({
+                    'username': username,
+                    'secret_hex': base64.b16encode(base64.b32decode(config['secret'])).decode()
+                })
+
+        return users
+
+    @private
+    async def get_ad_users(self):
+        return {
+            entry['user_sid']: entry for entry in await self.middleware.call(
+                'datastore.query', 'account.twofactor_user_auth', [['user_sid', '!=', None]]
+            )
+        }
+
+    @periodic(interval=86400, run_on_start=False)
+    @private
+    async def remove_expired_secrets(self):
+        if (await self.middleware.call('directoryservices.get_state'))['activedirectory'] != 'HEALTHY':
+            return
+
+        mapping = {
+            user['sid']: user for user in self.middleware.call_sync(
+                'user.query', [['local', '=', False], ['sid', '!=', None]], {
+                    'extra': {'additional_information': ['DS']},
+                }
             )
-        ]
+        }
+        for sid, entry in (await self.get_ad_users()).items():
```

Comment:
We can't rely on our cache to always be up-to-date. If you need to check whether a SID still maps to an AD user, then you can try `idmap.sid_to_name` call. libwbclient supports bulk conversion of sids to names. I can expose that API if you wish (then you can send it a large list of SIDs and only remove the ones that don't resolve).

Reply:
sid_to_name will raise `MatchNotFound` if it doesn't resolve.